Dockerfile

```diff
@@ -0,0 +1,116 @@
+# syntax=docker/dockerfile:1
+# This is loosely based on `docker init`'s rust template.
+# For a completely clean build, run something like this:
+# ```
+# docker build --build-arg=PROFILE=dev --build-arg=ENTRYPOINT=leader --no-cache
+# ```
+
+#############
+# Build stage
+#############
+# - `/src` is the repo directory.
+# - `/artifacts` is $CARGO_TARGET_DIR.
+# - `/output` is where the binaries go.
+
+ARG BUILD_BASE=rustlang/rust:nightly-bullseye-slim
+FROM ${BUILD_BASE} AS build
+
+# Install build dependencies.
+RUN apt-get update && apt-get install -y \
+    # for jemalloc
+    libjemalloc-dev \
+    libjemalloc2 \
+    make \
+    # for openssl
+    libssl-dev \
+    pkg-config \
+    # clean the image
+    && rm -rf /var/lib/apt/lists/*
+
+ARG PROFILE=release
+# forward the docker argument so that the script below can read it
+ENV PROFILE=${PROFILE}
+
+# Build the application.
+RUN \
```

Comment:
I think optimal way would be to mount host source and setup cache for build (instead of workspace member hacking/copying) in previous simple leader and worker dockerfiles, and skip this unified build/symbolic linking

Reply:
Are you advocating for not using cache mounts at all because they're too complicate[^1], or using them with our old dockerfiles? 

[^1]: https://github.com/0xPolygonZero/zk_evm/pull/357#discussion_r1663068377

Reply:
I advocate for using cache mount with old docker files. If you prefer, you can just select `leader` or `worker` with one ENV parameter but plain simple build.